### ISSUE TITLE
builtins.AttributeError: 'NoneType' object has no attribute 'writeHeaders'

### ISSUE THREAD

**Issue by nlamprian** [buildbot/buildbot#5979]
I have a step that adds multiple other steps when it runs. Occasionally, I get exceptions

```
2021-04-10 13:36:44+0200 [-] while handling API request
	Traceback (most recent call last):
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1475, in gotResult
	    _inlineCallbacks(r, g, status)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
	    result = g.send(result)
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 444, in renderRest
	    request.write(data)
	  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
	    self.gen.throw(type, value, traceback)
	--- <exception caught here> ---
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 135, in handleErrors
	    yield
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 444, in renderRest
	    request.write(data)
	  File "/usr/lib/python3/dist-packages/twisted/web/server.py", line 238, in write
	    http.Request.write(self, data)
	  File "/usr/lib/python3/dist-packages/twisted/web/http.py", line 1123, in write
	    self.channel.writeHeaders(version, code, reason, headers)
	builtins.AttributeError: 'NoneType' object has no attribute 'writeHeaders'
	
2021-04-10 13:36:44+0200 [-] While rendering resource:
	Traceback (most recent call last):
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1475, in gotResult
	    _inlineCallbacks(r, g, status)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1464, in _inlineCallbacks
	    status.deferred.errback()
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 501, in errback
	    self._startRunCallbacks(fail)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 568, in _startRunCallbacks
	    self._runCallbacks()
	--- <exception caught here> ---
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 654, in _runCallbacks
	    current.result = callback(current.result, *args, **kw)
	  File "/usr/lib/python3/dist-packages/buildbot/www/resource.py", line 96, in failHttpError
	    f.trap(Error)
	  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 439, in trap
	    self.raiseException()
	  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 467, in raiseException
	    raise self.value.with_traceback(self.tb)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 654, in _runCallbacks
	    current.result = callback(current.result, *args, **kw)
	  File "/usr/lib/python3/dist-packages/buildbot/www/resource.py", line 89, in failHttpRedirect
	    f.trap(Redirect)
	  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 439, in trap
	    self.raiseException()
	  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 467, in raiseException
	    raise self.value.with_traceback(self.tb)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
	    result = result.throwExceptionIntoGenerator(g)
	  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 491, in throwExceptionIntoGenerator
	    return g.throw(self.type, self.value, self.tb)
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 519, in asyncRender
	    res = yield self.renderRest(request)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
	    result = g.send(result)
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 444, in renderRest
	    request.write(data)
	  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
	    self.gen.throw(type, value, traceback)
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 165, in handleErrors
	    writeError(repr(e), errcode=500,
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 366, in writeError
	    request.write(data)
	  File "/usr/lib/python3/dist-packages/twisted/web/server.py", line 238, in write
	    http.Request.write(self, data)
	  File "/usr/lib/python3/dist-packages/twisted/web/http.py", line 1140, in write
	    self.channel.write(data)
	builtins.AttributeError: 'NoneType' object has no attribute 'write'
```

The problem is reproducible with

```python
class ManyStepsCommand(buildstep.ShellMixin, buildstep.BuildStep):

    def __init__(self, **kwargs):
        super().__init__(**self.setupShellMixin(kwargs))

    def run(self):
        self.build.addStepsAfterCurrentStep([steps.shell.ShellCommand(
            name=f"echo {i}", command=["echo", i]) for i in range(1, 30)])
        return results.SUCCESS
```

I'm using Buildbot 2.10.1.

Could you shed some light on this?

**Comment by p12tic** [comment]
Just looking into the stack trace this seems like it could be caused by a WWW client disconnection while a json response to some request is rendered. Could that be the case here?

**Comment by nlamprian** [comment]
Not sure how to interpret your comment. I haven't set up anything out of the ordinary.

I just happened to get another instance of the exception with a similar stacktrace (saw it only once so far). I forced a build from the website, ran the first step, and then this

```
2021-04-25 13:57:57+0200 [-] releaseLocks(ValidateRequest(name='validate request', haltOnFailure=True, description='validating', workdir='...')): []
2021-04-25 13:57:57+0200 [-]  step 'validate request' complete: success (None)
2021-04-25 13:57:57+0200 [-] while handling API request
	... same stuff ...
	  File "/usr/lib/python3/dist-packages/twisted/web/http.py", line 1138, in write
	    self.channel.writeSequence(toChunk(data))
	builtins.AttributeError: 'NoneType' object has no attribute 'writeSequence'
	
2021-04-25 13:57:57+0200 [-] added buildset 69 to database
2021-04-25 13:57:57+0200 [-] releaseLocks(TriggerTestBuild(set_properties={'owner': Property(owner), 'request': Property(got_request)})): []
2021-04-25 13:57:57+0200 [-]  step 'trigger' complete: success (None)
2021-04-25 13:57:57+0200 [-]  <Build request number:46 results:success>: build finished
```

So it's probably not connected to any particular step as suggested before. This time everything was on localhost.

**Comment by nlamprian** [comment]
Normally, this does not cause problems. But now, it had an instance of the exception after which the website became unresponsive (I'm assuming a connection here).

2021-05-04 15:27:20+0200 [-] releaseLocks(<Worker 'worker'>): []
2021-05-04 15:27:21+0200 [-] while handling API request
	Traceback (most recent call last):
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1475, in gotResult
	    _inlineCallbacks(r, g, status)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
	    result = g.send(result)
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 444, in renderRest
	    request.write(data)
	  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
	    self.gen.throw(type, value, traceback)
	--- <exception caught here> ---
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 135, in handleErrors
	    yield
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 444, in renderRest
	    request.write(data)
	  File "/usr/lib/python3/dist-packages/twisted/web/server.py", line 238, in write
	    http.Request.write(self, data)
	  File "/usr/lib/python3/dist-packages/twisted/web/http.py", line 1123, in write
	    self.channel.writeHeaders(version, code, reason, headers)
	builtins.AttributeError: 'NoneType' object has no attribute 'writeHeaders'
	
2021-05-04 15:27:21+0200 [-] While rendering resource:
	Traceback (most recent call last):
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1475, in gotResult
	    _inlineCallbacks(r, g, status)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1464, in _inlineCallbacks
	    status.deferred.errback()
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 501, in errback
	    self._startRunCallbacks(fail)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 568, in _startRunCallbacks
	    self._runCallbacks()
	--- <exception caught here> ---
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 654, in _runCallbacks
	    current.result = callback(current.result, *args, **kw)
	  File "/usr/lib/python3/dist-packages/buildbot/www/resource.py", line 96, in failHttpError
	    f.trap(Error)
	  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 439, in trap
	    self.raiseException()
	  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 467, in raiseException
	    raise self.value.with_traceback(self.tb)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 654, in _runCallbacks
	    current.result = callback(current.result, *args, **kw)
	  File "/usr/lib/python3/dist-packages/buildbot/www/resource.py", line 89, in failHttpRedirect
	    f.trap(Redirect)
	  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 439, in trap
	    self.raiseException()
	  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 467, in raiseException
	    raise self.value.with_traceback(self.tb)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
	    result = result.throwExceptionIntoGenerator(g)
	  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 491, in throwExceptionIntoGenerator
	    return g.throw(self.type, self.value, self.tb)
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 519, in asyncRender
	    res = yield self.renderRest(request)
	  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
	    result = g.send(result)
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 444, in renderRest
	    request.write(data)
	  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
	    self.gen.throw(type, value, traceback)
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 165, in handleErrors
	    writeError(repr(e), errcode=500,
	  File "/usr/lib/python3/dist-packages/buildbot/www/rest.py", line 366, in writeError
	    request.write(data)
	  File "/usr/lib/python3/dist-packages/twisted/web/server.py", line 238, in write
	    http.Request.write(self, data)
	  File "/usr/lib/python3/dist-packages/twisted/web/http.py", line 1140, in write
	    self.channel.write(data)
	builtins.AttributeError: 'NoneType' object has no attribute 'write'
	
2021-05-04 15:27:21+0200 [-] Sending mail (726 bytes) to ['xxx']

The server was sitting idle. If you have any more feedback to give me, I'd be happy to investigate further.
